server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java

```diff
@@ -0,0 +1,300 @@
+         * and leaves the door open for additional source states.
+         * @param shardNum an index into the shards which must be no greater than the number of shards before split
+         * @param sourceShardState the state to which the shard should be set
+         */
+        public void setSourceShardState(int shardNum, SourceShardState sourceShardState) {
+            assert shardNum >= 0 && shardNum < sourceShards.length : "source shardNum is out of bounds";
+            assert sourceShards[shardNum].ordinal() + 1 == sourceShardState.ordinal() : "invalid source shard state transition";
+            assert sourceShardState == SourceShardState.DONE : "can only move source shard state to DONE";
+            for (var target : getTargetStatesFor(shardNum)) {
+                assert target == TargetShardState.DONE : "can only move source shard to DONE when all targets are DONE";
+            }
+
+            sourceShards[shardNum] = sourceShardState;
+        }
+
+        /**
+         * Get the current target state of a shard
+         * @param shardNum an index into shards greater than or equal to the old shard count and less than the new shard count
+         * @return the target shard state for the shard identified by shardNum
+         */
+        public TargetShardState getTargetShardState(int shardNum) {
+            var targetShardNum = shardNum - oldShardCount;
+
+            assert targetShardNum >= 0 && targetShardNum < targetShards.length : "target shardNum is out of bounds";
+
+            return targetShards[targetShardNum];
+        }
+
+        /**
+         * Set the target state of a shard
+         * The only legal state in the split state machine is the one following the shard's current state.
+         * The reason for this API rather than an advanceState API is to confirm that the caller knows
+         * what the current state is when setting it.
+         * @param shardNum an index into shards greater than or equal to the old shard count and less than the new shard count
+         * @param targetShardState the state to which the shard should be set
+         */
+        public void setTargetShardState(int shardNum, TargetShardState targetShardState) {
+            var targetShardNum = shardNum - oldShardCount;
+
+            assert targetShardNum >= 0 && targetShardNum < targetShards.length : "target shardNum is out of bounds";
+            assert targetShards[targetShardNum].ordinal() + 1 == targetShardState.ordinal() : "invalid target shard state transition";
+
+            targetShards[targetShardNum] = targetShardState;
+        }
+
+        /**
+         * Get all the target shard states related to the given source shard
+         * @param shardNum a source shard index greater than or equal to 0 and less than the original shard count
+         * @return an array of target shard states in order for the given shard
+         */
+        public TargetShardState[] getTargetStatesFor(int shardNum) {
```

Comment:
I think we should either return a Collection of such states - or a Map<int, TargetShardState> to make the shard-id explicit.

I wonder if we even need to return this though or if we can add the methods needed to this class instead? Like the `inProgress` method below, but related to a specific source shard.

Reply:
I'll make this private and expose a `targetsDone(int sourceShardIdx)` or similar instead for the unit test.

Reply:
Done in 0e69a76